### ISSUE TITLE
Custom store for Nedb

### ISSUE THREAD

**Issue by dustintownsend** [diamondio/better-queue#13]
I needed a queue with nedb as the datastore, so I created a custom store for better-queue. 

Does this implementation look correct? I based it mostly on the SqlServerAdapter.

https://gist.github.com/dustintownsend/b701e5ffc620bb0c8bbde0d3f0bceefd

Edit: I forgot to note that I didn't fully implement priorities.

**Comment by leanderlee** [comment]
Looks like a good start! There are functionality missing, so you won't be able to take advantage of all of the features in the queue. When I run the test it fails 7 of the 23 tests.

I think the most concerning is that the `takeNextN` function isn't returning the tasks in order, so it's not going to act like a queue. Let me know if you run into trouble running the tests or implementing some parts.

**Comment by dustintownsend** [comment]
Thanks! For some reason I didn't think to run the test. I'll do that and see if I can figure out the issues. I also see the issue with takeNextN is I didn't add the sort for createdAt. 

Do you think creating a new repo & npm for this is a good idea? For example, better-queue-nedb? I'm new to node.js, javascript, and open source contributing. If my code didn't give that fact away already! I've seen other projects handle extensions in that way.

**Comment by dustintownsend** [comment]
@leanderlee which test where you running?

**Comment by leanderlee** [comment]
I was running `mocha test/basic.js` on the store... you might have to make a few changes to get it to point to the store.

I think a new repo would be good -- I'm hoping I'll get some time to write a different set of tests specifically for the stores.

**Comment by dustintownsend** [comment]
I went ahead and set-up a new repo for this https://github.com/dustintownsend/better-queue-nedb

And I imported the basic test and made some modifications for the test to use NedbStore. And I can now see the 7 test that are failing, so I can start working on getting those sorted out. Thanks for the help on this. 

I would like to make the code more similar to the default stores eventually. Now I'm returning the store as a property on NedbStore. The reason I did this was I was having an issue with getting the nedb datastore to persist across all calls to the methods. I know it's due to my lack of fully understanding how node modules actually work. Coming from C#, .NET, and Swift worlds doesn't seem to help me with that. 
I was also thinking it would be interesting to make this more generic to support MongoDB, as well. Since, Nedb is just using the MongoDB api. I was thinking about that when looking at how you had the adapter system setup for SQL Server.

**Comment by dustintownsend** [comment]
I believe I got the issue with takeNextN figured out. I was using createdAt (timestamp from nedb) for the ordering. I changed that to an added field that is sequential. I assumed that added was a timestamp. Thanks for pointing that out. 

Now down to 5 test failing. They all seem to be around timeout and cancel.

Is using _id instead of id going to cause any issues?

**Comment by leanderlee** [comment]
I don't believe the queue specifically checks for `id`, but you should make sure that the store returns the ID that was passed to it from `putTask`, and not a DB generated one.

**Comment by leanderlee** [comment]
@dustintownsend I've moved the better-queue stores into their own module, and added a test module so it'll be easy to test for compatibility with new stores in the future.

See:
https://github.com/diamondio/better-queue-sql
https://github.com/diamondio/better-queue-memory
https://github.com/diamondio/better-queue-store-test

**Comment by dustintownsend** [comment]
@leanderlee awesome! 

I had some issues with the nedbstore I created, so I decided to swap in your sqlite3 store for the time being. But, have been fighting issues getting sqlite3 and knex to build and package. For some reason knex breaks the installer builder. (I'm building an Electron app btw).

So, I'm going to take another pass at the nedbstore and see if I can get it passing all the test. I'll use your new better-queue-store-test that you set-up!
I also recently came across another javascript database called Lovefield I was considering trying as well for this.

**Comment by dustintownsend** [comment]
That was easy enough to get drop in the test. I got all the new test passing from better-queue-store-test. One failed at first and it helped me solve an issue with getting task! I was returning the entire result instead of just the task. That may have been the source of problems.

But, it is still failing 5 test from the original better-queue basic test. The timeout, cancel, type test.

The only thing I ran into with the new test was the "connect starts empty" test. I think it just returned faster than nedb/filesystem could create the file. So, if unlink failed in the destroy method, a 500ms timeout is called and tried again. Nothing major. In my case it wasn't cleaning up that one file.
